apstools/devices.py

```diff
@@ -906,27 +925,65 @@ class ApsUndulator(Device):
     """
 
-    energy = Component(EpicsSignal, "Energy", write_pv="EnergySet")
-    energy_taper = Component(EpicsSignal, "TaperEnergy", write_pv="TaperEnergySet")
+    energy = Component(EpicsSignal, "Energy", write_pv="EnergySet",
+                       put_complete=True, kind='hinted')
+    energy_taper = Component(EpicsSignal, "TaperEnergy",
+                             write_pv="TaperEnergySet", kind='config')
     gap = Component(EpicsSignal, "Gap", write_pv="GapSet")
-    gap_taper = Component(EpicsSignal, "TaperGap", write_pv="TaperGapSet")
-    start_button = Component(EpicsSignal, "Start", put_complete=True)
-    stop_button = Component(EpicsSignal, "Stop")
-    harmonic_value = Component(EpicsSignal, "HarmonicValue")
-    gap_deadband = Component(EpicsSignal, "DeadbandGap")
-    device_limit = Component(EpicsSignal, "DeviceLimit")
-
-    access_mode = Component(EpicsSignalRO, "AccessSecurity")
-    device_status = Component(EpicsSignalRO, "Busy")
-    total_power = Component(EpicsSignalRO, "TotalPower")
-    message1 = Component(EpicsSignalRO, "Message1")
-    message2 = Component(EpicsSignalRO, "Message2")
-    message3 = Component(EpicsSignalRO, "Message3")
-    time_left = Component(EpicsSignalRO, "ShClosedTime")
-
-    device = Component(EpicsSignalRO, "Device")
-    location = Component(EpicsSignalRO, "Location")
-    version = Component(EpicsSignalRO, "Version")
+    gap_taper = Component(EpicsSignal, "TaperGap", write_pv="TaperGapSet",
+                          kind='config')
+    start_button = Component(EpicsSignal, "Start", put_complete=True,
+                             kind='omitted')
+    stop_button = Component(EpicsSignal, "Stop", kind='omitted')
+    harmonic_value = Component(EpicsSignal, "HarmonicValue", kind='config')
+    gap_deadband = Component(EpicsSignal, "DeadbandGap", kind='config')
+    device_limit = Component(EpicsSignal, "DeviceLimit", kind='config')
+
+    access_mode = Component(EpicsSignalRO, "AccessSecurity", kind='omitted')
+    device_status = Component(EpicsSignalRO, "Busy", kind='omitted')
+    total_power = Component(EpicsSignalRO, "TotalPower", kind='config')
+    message1 = Component(EpicsSignalRO, "Message1", kind='omitted')
+    message2 = Component(EpicsSignalRO, "Message2", kind='omitted')
+    message3 = Component(EpicsSignalRO, "Message3", kind='omitted')
+    time_left = Component(EpicsSignalRO, "ShClosedTime", kind='omitted')
+
+    device = Component(EpicsSignalRO, "Device", kind='config')
+    location = Component(EpicsSignalRO, "Location", kind='config')
+    version = Component(EpicsSignalRO, "Version", kind='config')
+
+    # Useful undulator parameters that are not EPICS PVs.
+    energy_deadband = Component(Signal, value=0.0, kind='config')
+    energy_backlash = Component(Signal, value=0.0, kind='config')
+    energy_offset = Component(Signal, value=0, kind='config')
+    tracking = Component(TrackingSignal, value=False, kind='config')
+
+    def undulator_setup(self):
```

Comment:
I'd rather not add any code to `apstools` that involves command line Q&A such as this.  Can the same function be replaced with [getter/setter decorators](https://stackabuse.com/the-python-property-decorator/)?

Reply:
I don't think this function could be replaced with getter/setter. I'll remove it.